File: java/com/lc/lc108.java
package lc;

import java.util.*;

public class lc108 {

    public static void main(String[] args) {
        new lc108().sortedArrayToBST(new int[] { -10, -3, 0, 5, 9 });
    }

    public TreeNode sortedArrayToBST(int[] nums) {
        return dfs(nums, 0, nums.length);
    }

    public TreeNode dfs(int[] nums, int l, int r) {
        int m = (l + r) >> 1;
        if (l >= r || m < 0 || m >= nums.length) {
            return null;
        }
        TreeNode cur = new TreeNode(nums[m]);
        cur.left = dfs(nums, l, m);
        cur.right = dfs(nums, m + 1, r);
        return cur;
    }
}

File: go/lc/lc108.go
package main

func sortedArrayToBST(nums []int) *TreeNode {
	var dfs func(nums []int, l, r int) *TreeNode
	dfs = func(nums []int, l, r int) *TreeNode {
		if l > r {
			return nil
		}
		mid := ((l + r) >> 1)
		cur := &TreeNode{Val: nums[mid]}
		cur.Left = dfs(nums, l, mid-1)
		cur.Right = dfs(nums, mid+1, r)
		return cur
	}

	return dfs(nums, 0, len(nums)-1)
}
